### PR TITLE
util: Move route table calculator from egressip

### PR DESCRIPTION
**Pull request by qinqon**
#### What this PR does and why is it needed

A new vrf table id is needed by UDN, this change move the egress ip table ID generator to util so it can be used by UDN feature.

Also a new command to `ovnkube-trace --dump-udn-vrf-table-ids` to dump VRF table IDs per node, the output is a json in the form 
```json
{
"ovn-control-plane": {
   "net-blue": 1,
   "net-red": 2
},
"ovn-worker": {
   "net-blue": 3,
   "net-red": 4
}
}
```

#### How to verify it
The egress ip feature already verify the util

#### Description for the changelog
util: Move route table calculator from egressip

#### Does this PR introduce a user-facing change?
This change introduces a new flag at `ovnkube-trace --dump-udn-vrf-table-ids` 

```release-note
NONE
```

### CODE REVIEW

**Review by tssurya — COMMENTED** [review]
commit1 is lgtm except 1 small nit
commit2 asked a qq on the design

File: go-controller/pkg/util/net.go
Comment (qinqon) [reply go-controller/pkg/util/net.go:332]: Just saying "interface" means nothing, this is the index of a network interface.

Comment (tssurya) [reply go-controller/pkg/util/net.go:332]: This file in total os called net.go but yeah disregard this comment it was a nit :)

File: go-controller/cmd/ovnkube-trace/ovnkube-trace.go
Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1182]: > nit: its too long (the name maybe just `dumpUDNVRFIDs`) ? for network is understood since you have UDN also in future if we want to do the same for finding ovn allocated zones, marks, we might want to make this more extensible?
> 

I will rename into `dumpVRFTableIDs` 

> so if we go with `--dump-udn-vrf-table-ids` it dumps the map right?: networka: 10 networkb: 15
> 
> like that? or are you taking network name as input and then dumping the vrfid for that network across all nodes in the cluster? could you please add some more details to your commit message?

VRF Table IDs are going to be different per Node since the interface index is something decided by the kernel, so this is going to return `{node1: 10 node2: 15}` for the network specified

I am going to just print all the networks and move the flag as a boolean.

Comment (tssurya) [reply go-controller/cmd/ovnkube-trace/ovnkube-trace.go]: yes I like that better, let's print all networks! also let's call out that this command is expected to output the results of a specific node right? its not that we plan to take nodeName as a parameter?

Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/ovnkube-trace.go]: It will print it for all the nodes, users can filter by node the output using `jq` or the like.

File: go-controller/cmd/ovnkube-trace/ovnkube-trace.go
Comment (tssurya) [inline go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1182]: add a new docs section similar to https://ovn-kubernetes.io/troubleshooting/ovnkube-trace but a totally new udn.md file under troubleshooting and start with this command flag explanation along with a sample output

File: go-controller/cmd/ovnkube-trace/ovnkube-trace.go
Comment (tssurya) [inline go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1189]: why did we move these checks to a new place?

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (tssurya) [inline go-controller/cmd/ovnkube-trace/udn.go:45]: this will work on both ovnkube-node and ovnkube-control-plane pods? which one will you pick on a master node?

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (tssurya) [inline go-controller/cmd/ovnkube-trace/udn.go:51]: ip is installed by default on ovnkube-node pod right? which is host-networked?

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (tssurya) [inline go-controller/cmd/ovnkube-trace/udn.go:79]: can you please show me a sample output on a cluster with 2 or 3 UDNs?

**Review by tssurya — CHANGES_REQUESTED** [review]

File: go-controller/cmd/ovnkube-trace/ovnkube-trace.go
Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1189]: Retrieving the kubeconfig and coreclient is now done before these checks since it's needed fo discover the tableIDs, laeter on when we refactor with subcommands it will look better.

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/udn.go:45]: ovnkube-trace select only the ovnkube-node pods and there you have ovnkube-controller.

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/udn.go:51]: yep.

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/udn.go:79]:
```json
{
"ovn-control-plane": {
   "net-blue": 1,
   "net-red": 2
},
"ovn-worker": {
   "net-blue": 3,
   "net-red": 4
}
}
```

File: go-controller/cmd/ovnkube-trace/ovnkube-trace.go
Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1182]: Done

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (tssurya) [reply go-controller/cmd/ovnkube-trace/udn.go:79]: awesome looks pretty thanks!

File: go-controller/cmd/ovnkube-trace/ovnkube-trace.go
Comment (martinkennelly) [inline go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1235]: nit: two error messages are the same so would be hard to tell which path we hit.

**Review by martinkennelly — COMMENTED** [review]
small nit otherwise LGTM

File: go-controller/cmd/ovnkube-trace/ovnkube-trace.go
Comment (martinkennelly) [reply go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1235]: another nit but its fine for this PR; i am a fan of adding context to error messages (if needed) to allow the user introspect where it failed, otherwise theyre more likely to need dev support.

Comment (qinqon) [reply go-controller/cmd/ovnkube-trace/ovnkube-trace.go:1235]: Ahh wait, this is from an old version, since we are already returning string with JSON, I will remove the Marshaling.

**Review by tssurya — APPROVED** [review]
I am lgtm here
waiting for @martinkennelly 's and @pperiyasamy 's review before merging.

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (tssurya) [reply go-controller/cmd/ovnkube-trace/udn.go:45]: @qinqon: for nonIC it will pick ovnkube-node pod I hope

**Review by martinkennelly — APPROVED** [review]

File: go-controller/cmd/ovnkube-trace/udn.go
Comment (pperiyasamy) [inline go-controller/cmd/ovnkube-trace/udn.go:50]: nit. we may move this  as an util like `getNetMgmtPortLinkName()` so that others can also use it. but let's do this in some other PR.

**Review by pperiyasamy — APPROVED** [review]

**Comment by tssurya** [comment]
@qinqon : why am I getting :+1: 
```
$ ./_output/go/bin/ovnkube-trace -dump-udn-vrf-table-ids
F0703 18:53:53.418323  184763 ovnkube-trace.go:1231] Failed dumping VRF table IDs: : Device "default_ovn-k8s-mp0" does not exist.
: command terminated with exit code 1
```
on master? :D

**Comment by qinqon** [comment]
> (probably pulling in the changes a bit too late :) ...) @qinqon : why am I getting this:
> 
> ```
> 
> $ ./_output/go/bin/ovnkube-trace -dump-udn-vrf-table-ids
> 
> F0703 18:53:53.418323  184763 ovnkube-trace.go:1231] Failed dumping VRF table IDs: : Device "default_ovn-k8s-mp0" does not exist.
> 
> : command terminated with exit code 1
> 
> ```
> 
> on master? :D default is not prefixed to mp0 for default network hehe :D

We should skip default, will put a fix PR

**Comment by tssurya** [comment]
https://github.com/ovn-org/ovn-kubernetes/pull/4498 fixes the problem seen here

**Pull request merged**
